brave-kafka/src/main/java/smartthings/brave/kafka/consumers/TracingConsumerInterceptor.java

```diff
@@ -0,0 +1,128 @@
+
+  private Tracer tracer;
+  private Endpoint kafkaEndpoint;
+
+  // TODO: how to inject dependencies without using config
+//  private final SpanNameProvider<K> nameProvider;
+
+  public TracingConsumerInterceptor() {}
+
+  private final TraceContext.Extractor<EnvelopeProtos.Envelope> extractor = new TraceContext.Extractor<EnvelopeProtos.Envelope>() {
+    @Override
+    public TraceContextOrSamplingFlags extract(EnvelopeProtos.Envelope envelope) {
+      TraceContext.Builder builder = TraceContext.newBuilder()
+        .traceIdHigh(envelope.getTraceIdHigh())
+        .traceId(envelope.getTraceId())
+        .spanId(envelope.getSpanId())
+        .sampled(envelope.getSampled())
+        .shared(envelope.getShared());
+      if (envelope.getParentId() != 0) {
+        builder.parentId(envelope.getParentId());
+      }
+      return TraceContextOrSamplingFlags.create(builder);
+    }
+  };
+
+
+  @Override
+  public ConsumerRecords<K, byte[]> onConsume(ConsumerRecords<K, byte[]> records) {
+    Map<TopicPartition, List<ConsumerRecord<K, byte[]>>> unenvelopedRecords = new HashMap<>(records.count());
+    records.forEach(record -> {
+      try {
+        EnvelopeProtos.Envelope envelope = EnvelopeProtos.Envelope.parseFrom(record.value());
+        TraceContextOrSamplingFlags traceContextOrSamplingFlags = extractor.extract(envelope);
+        TraceContext ctx = traceContextOrSamplingFlags.context();
+        (ctx != null ? tracer.joinSpan(ctx) : tracer.newTrace(traceContextOrSamplingFlags.samplingFlags()))
+          .name("Receive " + record.topic() + "-" + record.partition())
+          .annotate(Constants.SERVER_RECV)
+          .remoteEndpoint(kafkaEndpoint)
+          .flush();
+        // Some of these seem wrong (checksum, value size) but we need to set the timestamp using this constructor
+        ConsumerRecord<K, byte[]> unenvelopedRecord = new ConsumerRecord<>(record.topic(), record.partition(), record.offset(),
+          record.timestamp(), record.timestampType(), record.checksum(), record.serializedKeySize(),
+          record.serializedValueSize(), record.key(), envelope.getPayload().toByteArray());
+
+        TopicPartition tp = new TopicPartition(record.topic(), record.partition());
+        if (unenvelopedRecords.containsKey(tp)) {
+          unenvelopedRecords.get(tp).add(unenvelopedRecord);
+        } else {
+          List<ConsumerRecord<K, byte[]>> list = new LinkedList<ConsumerRecord<K, byte[]>>();
+          list.add(unenvelopedRecord);
+          unenvelopedRecords.put(tp, list);
+        }
+      } catch (InvalidProtocolBufferException e) {
+        e.printStackTrace();
+      }
+    });
+    return new ConsumerRecords<>(unenvelopedRecords);
+  }
+
+  @Override
+  public void onCommit(Map<TopicPartition, OffsetAndMetadata> offsets) {
+
+  }
+
+  @Override
+  public void close() {
+
+  }
+
+  @Override
+  public void configure(Map<String, ?> configs) {
```

Comment:
Configuration/construction is often better described though a builder pattern.  For an example:
https://github.com/openzipkin/brave/blob/master/brave/src/main/java/brave/Tracer.java#L50